ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -127,65 +129,103 @@ public SchemaKStream filter(final Expression filterExpression) {
     );
   }
 
-  public SchemaKStream select(final Schema selectSchema) {
-    final KStream<String, GenericRow>
-        projectedKStream =
-        kstream.mapValues(row -> {
-          List<Object> newColumns = new ArrayList<>();
-          for (Field schemaField : selectSchema.fields()) {
-            newColumns.add(extractColumn(schemaField, row)
-            );
-          }
-          return new GenericRow(newColumns);
-        });
-
+  public SchemaKStream select(final List<Pair<String, Expression>> expressionPairList) {
+    Selection selection = new Selection(expressionPairList, functionRegistry, this);
     return new SchemaKStream(
-        selectSchema,
-        projectedKStream,
-        keyField,
+        selection.getSchema(),
+        kstream.mapValues(selection.getSelectValueMapper()),
+        selection.getKey(),
         Collections.singletonList(this),
         Type.PROJECT,
         functionRegistry,
         schemaRegistryClient
     );
   }
 
-  public SchemaKStream select(final List<Pair<String, Expression>> expressionPairList) {
-    final Pair<Schema, SelectValueMapper> schemaAndMapper = createSelectValueMapperAndSchema(
-        expressionPairList);
+  static class Selection {
+    private Schema schema;
+    private Field key;
+    private SelectValueMapper selectValueMapper;
+
+    public Selection(
+        final List<Pair<String, Expression>> expressionPairList,
+        final FunctionRegistry functionRegistry,
+        final SchemaKStream fromStream) {
+      key = findKeyField(expressionPairList, fromStream);
+      List<ExpressionMetadata> expressionEvaluators = buildExpressions(
+          expressionPairList, functionRegistry, fromStream);
+      schema = buildSchema(expressionPairList, expressionEvaluators);
+      selectValueMapper = new SelectValueMapper(
+          new GenericRowValueTypeEnforcer(
+              fromStream.getSchema()), expressionPairList, expressionEvaluators);
+    }
 
-    return new SchemaKStream(
-        schemaAndMapper.left,
-        kstream.mapValues(schemaAndMapper.right),
-        keyField,
-        Collections.singletonList(this),
-        Type.PROJECT,
-        functionRegistry,
-        schemaRegistryClient
-    );
-  }
+    private Field findKeyField(
```

Comment:
Why introduce this method? Is necessary for `SchemaKTable` and nice to have for `SchemaKStream`?

Reply:
This change encapsulates management of the key field away from the client code. So if the select call renames the key field then the SchemaKStream/SchemaKTable figures out the new key field and sets it correctly in the returned SchemaKStream/SchemaKTable.

This way AggregateNode doesn't need the maybeSetTheKeyField stuff that creates a new SchemaKStream with the key field that uses internal names. I think in general we should try and avoid constructing these SchemaKStream and SchemaKTable objects from plan nodes (maybe even going as far as making the constructors package-private and providing a static factory function for just StructuredDataSourceNode to use). This avoids the problem of accidentally converting from a SchemaKTable to a SchemaKStream.